valkyrie/lib/valkyrie/storage_adapter.rb

```diff
@@ -36,6 +36,73 @@ def self.find_by(id:)
       end.find_by(id: id)
     end
 
+    class Fedora
+      attr_reader :connection
+      PROTOCOL = 'fedora://'
+      def initialize(connection:)
+        @connection = connection
+      end
+
+      # @param id [Valkyrie::ID]
+      # @return [Boolean] true if this adapter can handle this type of identifer
+      def handles?(id:)
+        id.to_s.start_with?(PROTOCOL)
+      end
+
+      # Return the file associated with the given identifier
+      # @param id [Valkyrie::ID]
+      # @return [Valkyrie::StorageAdapter::File]
+      def find_by(id:)
+        File.new(id: id, io: response(id: id))
+      end
+
+      # @param file [IO]
+      # @param model [Valkyrie::Model]
+      # @return [Valkyrie::StorageAdapter::File]
+      def upload(file:, model:)
+        # TODO: this is a very naive aproach. Change to PCDM
+        identifier = model.id.to_uri + '/original'
+        ActiveFedora::File.new(identifier) do |af|
+          af.content = file
+          af.save!
+          af.metadata.set_value(:type, af.metadata.type + [::RDF::URI('http://pcdm.org/use#OriginalFile')])
+          af.metadata.save
+        end
+        find_by(id: Valkyrie::ID.new(identifier.to_s.sub(/^.+\/\//, PROTOCOL)))
+      end
+
+      class IOProxy
+        # @param faraday_response [Faraday::Response]
+        def initialize(faraday_response)
+          @faraday_response = faraday_response
+        end
+        delegate :read, to: :io
+
+        # There is no streaming support in faraday (https://github.com/lostisland/faraday/pull/604)
```

Comment:
😞 

Outside the scope of this ticket, so I've opened a separate ticket to support streaming binary content out of Fedora: #145